### PR TITLE
Issue #5304 HTTP2 HostHeader

### LINKED ISSUE #5304: HTTP/2 with HttpServletRequest.getHeader("Host") returns null on Jetty 10, but a valid value on Jetty 9
**Jetty version**
10.0.x

**Description**
If you make an HTTP/2 request, and then in your servlet you use `HttpServletRequest.getHeader("Host")` the value is null on Jetty 10.
But it returns a value on Jetty 9.

**Screenshot of Jetty 9** (notice the "Host" header in the enumerated headers)

jetty-9-chrome-http2-authority-header

**Screenshot of Jetty 10** (no "Host" header in the enumerated headers)

jetty-10-chrome-http2-authority-header

### PR DESCRIPTION
**Pull request by gregw**
#5304 
 Updated HostHeaderCustomizer to actually add the Host header, either from values passed in the custructor or from the getServerName and getServerPort methods.

The HttpURI is no longer updated.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java [hunk 1]
```diff
@@ -19,30 +19,30 @@
...
+ * The host and port may be provided in the constructor or taken from the
+ * {@link Request#getServerName()} and {@link Request#getServerPort()} methods.
+ * </p>
  */
 public class HostHeaderCustomizer implements HttpConfiguration.Customizer
 {
     private final String serverName;
     private final int serverPort;
 
+    public HostHeaderCustomizer()
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:44]: Some javadoc here saying this usage will result in using the serverName/serverPort from the Request?

File: jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java [hunk 2]
```diff
@@ -19,30 +19,30 @@
...
+    {
+        this(null, 0);
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:46]: This forwards to `this(String, int)` which requires that `serverName` is not null.

**Review by joakime — CHANGES_REQUESTED** [review]
This probably needs test cases, as there is a flaw in the null serverName case in this code.

File: jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java [hunk 3]
```diff
@@ -57,15 +60,26 @@ public HostHeaderCustomizer(String serverName)
...
     {
-        if (request.getHeader("Host") == null)
-            // TODO set the field as well?
-            request.setHttpURI(HttpURI.build(request.getHttpURI()).host(serverName).port(serverPort));
+        if (request.getHttpVersion() != HttpVersion.HTTP_1_1 && !request.getHttpFields().contains(HttpHeader.HOST))
+        {
+            String host = serverName == null ? request.getServerName() : serverName;
+            int port = HttpScheme.normalizePort(request.getScheme(), serverPort == 0 ? request.getServerPort() : serverPort);
+
+            if (serverName != null || serverPort > 0)
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:75]: Don't you want to setHttpURI() if the `port` is normalized in a way that's different then `serverPort` too?

File: jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java [hunk 4]
```diff
@@ -57,15 +60,26 @@ public HostHeaderCustomizer(String serverName)
...
+                request.setHttpURI(HttpURI.build(request.getHttpURI()).authority(host, port));
+
+            HttpFields original = request.getHttpFields();
+            HttpFields.Mutable httpFields = HttpFields.build(original.size() + 1);
+            httpFields.add(new HostPortHttpField(host, port));
+            httpFields.add(request.getHttpFields());
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:81]: This looks odd, why is this not `httpFields.addAll(request.getHttpFields());` ?

Also, doesn't this change the `request.getHttpFields()` to now be Mutable?

What if this was ...

``` java
HttpFields updated = HttpFields.build(
   request.getHttpFields(), 
   new HostPortHttpFields(host, port));
request.setHttpFields(updated);
```

Where there is a new method `.build(HttpFields original, HttpField ... fields)` ?

**Review by joakime — CHANGES_REQUESTED** [review]
Some minor concerns.

File: jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java (hunk 4 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:81]: https://tools.ietf.org/html/rfc7230#section-3.2.2 says Host header should be one of the first headers sent.  I can't imagine too much depending on this, but best to follow recommendations.

File: jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java (hunk 3 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:75]: My thinking here is that if the serverPort is set, then we will set the port, but will used the normalised version of it.
So if they set port 80, but the real port was 8888 we would set the port to 0 not 80

**Comment by gregw** [comment]
@joakime nudge

File: jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java (hunk 4 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:81]: This still changes future requests on `request.getHttpFields()` to be a `HttpFields.Mutable`, right?
I don't think that's desired.

**Review by joakime — CHANGES_REQUESTED** [review]
accidentally Mutable?

**Comment by gregw** [comment]
@joakim, it's not accidentally mutable. The style we mostly use is that the setter decides if it wants the mutable or immutable reference.  So in this case request will call asImutable on the passed HttpField.

The idea is that the callee decides and the caller can never make the request fields mutable

File: jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java (hunk 4 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:81]: https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java#L250-L253

**Review by joakime — APPROVED** [review]
LGTM
